### DISCUSSION TITLE
Two's complement bug in time.mktime()

### DISCUSSION THREAD

**Discussion by c-max** [micropython/micropython#15600] [General]
Hello again

time.mktime() seems to have a two's complement problem.
It returns the 32bit unsigned value instead of the 32bit signed value. 

Got this problem with the latest release and the latest preview on an esp32.

(For further questions: 
How do I know if a bug/problem is general or only esp32 if the esp32 is my only testing hardware?)

Code:

```py
import os,time

for e in str(os.uname())[1:-1].split(','):
    print(e.strip())

timestamp_orig = -1
gmt = time.gmtime(timestamp_orig)
print(f'\n{gmt}\n')

timestamp_comp = time.mktime(gmt)
print (f'{timestamp_orig} {timestamp_comp}')

print (f"^^^ that's an error itself and leads to further errors:")
gmt = time.gmtime(timestamp_comp)
```

Output:

```
sysname='esp32'
nodename='esp32'
release='1.24.0-preview'
version='v1.24.0-preview.163.gad3552ae2 on 2024-08-02'
machine='Generic ESP32 module with ESP32'

(1999, 12, 31, 23, 59, 59, 4, 365)

-1 4294967295
^^^ that's an error itself and leads to further errors:
Traceback (most recent call last):
  File "<stdin>", line 18, in <module>
OverflowError: overflow converting long int to machine word
```

Have a nice week
c-max

**Comment by peterhinch** [comment]
My understanding is that, for reasons of efficiency on 32 bit hardware, times prior to the epoch are not supported.

**Reply by c-max** [reply]
Sounds reasonable for practical use.
Testing boundary values it confused me.

**Comment by krzotr** [comment]
> How do I know if a bug/problem is general or only esp32 if the esp32 is my only testing hardware?)

Have a look [ESP32 - Exercise 8: RTC and data logging - Exercise 6: Data logging](https://afnog.iotworkshop.africa/do/view/IoT_Course_English/RealTimeClockAndDataLogging)

This error `OverflowError: overflow converting long int to machine word` also occurs on my `LILYGO TTGO LoRa32 with ESP32 - MicroPython 91f4a6b on 2024-08-09;`

> Concerning the time stamp there is a difference between MicroPython and CPython: MicroPython starts counting at **2000-01-01** while CPython uses Unix epoch, which starts at **1970-01-01**.
>
> timeStampCPython = timeStampMicroPython + time.UNIX_TIME which is the constant **946677600**
> 
> MPY epoch is **2000-01-01** instead of unix epoch, which is 1970-01-01.

**Reply by scruss** [reply]
> … MicroPython starts counting at 2000-01-01 

is not quite right. ESP32 starts at 2000, rp2 and others do not:

```
sysname='rp2'
nodename='rp2'
release='1.23.0'
version='v1.23.0 on 2024-06-02 (GNU 13.2.0 MinSizeRel)'
machine='Raspberry Pi Pico with RP2040'

(1969, 12, 31, 23, 59, 59, 2, 365)

-1 4294967295
^^^ that's an error itself and leads to further errors:
Traceback (most recent call last):
  File "<stdin>", line 14, in <module>
OverflowError: overflow converting long int to machine word
```
No matter what the epoch, negative offsets aren't supported on 32 bits
